pypose/optim/controller_parameters_tuner.py

```diff
@@ -0,0 +1,101 @@
+
+    def tune(self, dynamic_system, initial_state, ref_states, controller, parameters,
+             parameters_tuning_set, tau, states_to_tune, func_get_state_error):
+        r"""
+        Args:
+            dynamic_system (pypose.module.dynamics):
+            initial_state (Tensor):
+            ref_states ():
+            controller (pypose.module.controller): Linear or nonlinear controller to control
+            the dynamic system
+            parameters (Tensor): Controller parameters
+            parameters_tuning_set (Tensor): This set gives the minimum and the maximum
+                value the parameters can reach
+            tau: time interval in system
+            states_to_tune (Tensor): choose which state needs to be considered in the
+                loss function, usually only position is chosen.
+            func_get_state_error (function): function has two inputs: system state and
+                ref state. The function needs to be provided by users considering system
+                state and ref state are not always in the same formation.
+        """
+        states_to_tune = states_to_tune.double()
+        states = []
+        dxdparam_gradients = []
+
+        system_state = torch.clone(initial_state)
+        controller_parameters = parameters
+        states.append(system_state)
+        dxdparam_gradients.append(
+            torch.zeros(
+                [len(initial_state), len(controller_parameters)], device=self.device)
+            .double())
+
+        for index, ref_state in enumerate(ref_states):
+            controller_input = controller.get_control(parameters=controller_parameters, state=system_state, ref_state=ref_state, feed_forward_quantity=None)
+            system_new_state = dynamic_system.state_transition(system_state, controller_input, tau)
+
+            # calcuate the derivative
+            dhdx_func = lambda state: controller.get_control(parameters = controller_parameters, state = state, ref_state = ref_state, feed_forward_quantity = None)
+            dhdxk_tensor = torch.squeeze(jacobian(dhdx_func, system_state))
+
+            dhdparam_func = lambda params: controller.get_control(parameters = params, state = system_state, ref_state = ref_state, feed_forward_quantity = None)
+            dhdparam_tensor = torch.squeeze(jacobian(dhdparam_func, controller_parameters))
+
+            dfdxk_func = lambda system_state: dynamic_system.state_transition(state = system_state, input = controller_input, t = tau)
+            dfdxk_tensor = torch.squeeze(jacobian(dfdxk_func, system_state))
+
+            dfduk_func = lambda inputs: dynamic_system.state_transition(state = system_state, input = inputs, t = tau)
+            dfduk_tensor = torch.squeeze(jacobian(dfduk_func, controller_input))
+
+            states.append(system_new_state)
+            system_state = system_new_state
+
+            last_gradient = dxdparam_gradients[-1]
+
+            dxdparam_gradients.append(
```

Comment:
Add the partial derivative with respect of u_k to accommodate for the general Loss function. such as the one with control penalty term.